hazelcast/src/main/java/com/hazelcast/map/impl/MapSplitBrainHandlerService.java

```diff
@@ -20,331 +20,240 @@
     }
 
     @Override
     public Runnable prepareMergeRunnable() {
-        long now = Clock.currentTimeMillis();
-
-        Map<String, MapContainer> mapContainers = mapServiceContext.getMapContainers();
-        Map<MapContainer, Collection<Record>> recordMap = createHashMap(mapContainers.size());
-
-        for (MapContainer mapContainer : mapContainers.values()) {
-            MapConfig mapConfig = mapContainer.getMapConfig();
-            InMemoryFormat inMemoryFormat = mapConfig.getInMemoryFormat();
-            if (inMemoryFormat == NATIVE
-                    && nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
-                logger.warning("Split-brain recovery can not be applied NATIVE in-memory-formatted map ["
-                        + mapContainer.name + ']');
-
-                continue;
-            }
-
-            Object mergePolicy = getMergePolicy(mapConfig.getMergePolicyConfig());
-            boolean mergePartitionData = !(mergePolicy instanceof IgnoreMergingEntryMapMergePolicy)
-                    && !(mergePolicy instanceof DiscardMergePolicy);
+        Map<String, Collection<RecordStore>> collectedStores = new HashMap<String, Collection<RecordStore>>();
+        Map<String, Collection<RecordStore>> collectedStoresWithLegacyPolicies
+                = new HashMap<String, Collection<RecordStore>>();
+        List<RecordStore> backupStores = new LinkedList<RecordStore>();
+
+        PartitionContainer[] partitionContainers = mapServiceContext.getPartitionContainers();
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+            boolean partitionOwner = partitionService.isPartitionOwner(partitionId);
+
+            List<Iterator<RecordStore>> iterators = iteratorsOf(partitionContainers[partitionId]);
+            for (Iterator<RecordStore> iterator : iterators) {
+                while (iterator.hasNext()) {
+                    RecordStore store = iterator.next();
+                    Object mergePolicy = getMergePolicy(store.getMapContainer().getMapConfig().getMergePolicyConfig());
+
+                    if (isIgnoreOrDiscardPolicy(mergePolicy)) {
+                        continue;
+                    }
 
-            for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
-                RecordStore<Record> recordStore = getOrNullRecordStore(mapContainer.name, inMemoryFormat, partitionId);
-                if (recordStore == null) {
-                    continue;
-                }
-                // add your owned entries to the map so they will be merged
-                if (mergePartitionData && partitionService.isPartitionOwner(partitionId)) {
-                    Collection<Record> records = recordMap.get(mapContainer);
-                    if (records == null) {
-                        records = new LinkedList<Record>();
-                        recordMap.put(mapContainer, records);
+                    if (!partitionOwner) {
```

Comment:
The `isIgnoreOrDicardPolicy` check is done before this. I eventually see `destroyStores(backupStores);` in the `AbstractMerger`. Could this cause an HD memory leak, since we skip some `backupStores` and don't destroy them?

Reply:
good one. fixed